arcconfig/arcconfig.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package arcconfig defines the API Runtime Control config and provides
+// the config loading and validation functions.
+package arcconfig
+
+import "time"
+
+// APIRuntimeControlConfig is an Apigee Environment-level config for
+// Envoy Adapter. It contains a list of operations for the adapter to
+// perform request authentication and authorization.
+type APIRuntimeControlConfig struct {
+	// Name of the config
+	Name string `yaml:"name"`
+
+	// Revision of the config
+	Revision string `yaml:"revision"`
+
+	// A list of Operations, names of which must be unique within the config.
+	Operations []*Operation `yaml:"operations,omitempty"`
+}
+
+// An API Operation associates a set of rules with a set of request matching
+// settings.
+type Operation struct {
+	// Name of the operation. Unique within a APIRuntimeControlConfig.
+	Name string `yaml:"name"`
+
+	// Authentication defines the AuthenticationRequirement
+	Authentication AuthenticationRequirement `yaml:"authentication,omitempty"`
+
+	// Authorization defines the ConsumerAuthorization
+	Authorization ConsumerAuthorization `yaml:"authorization,omitempty"`
+
+	// HTTP matching rules for this operation.
+	// The path templates within which should be fully denormalized.
```

Comment:
Is this implying it includes the basepath? We decided last week that we'd keep basepath at a higher level i.e. the first layer of the tree.

Reply:
Yes I think the basepath should be kept at a higher level for the other schema. I just thought when we generate this config from it, we could denormalize all the paths here so that ARC/EA doesn't have to compute this. Also we need to do the denormalization in control plane anyway for the ingress contract generation.

Reply:
Yeah... probably we should keep this at the higher level. That matches Open API's approach as well. Computing the full path is trivial, but requiring users to enter it repeatedly is lame. :)

Reply:
Makes sense. Will add a top one.